### ISSUE TITLE
Consistent error seen when attempting to upload to JDB

### ISSUE THREAD

**Issue by rwaldron** [Jewelbots/arduino-library#3]
I'm filing this separately because I'm not sure if it's a dupe of #1. If it is, then just close this.

> Failed to execute script __main__
Upgrading target on /dev/cu.usbserial-DB00W8D4 with DFU package /private/var/folders/xz/k2lxgs3s43180_tjwn1v_kwc0000gn/T/arduino_build_939294/jewelbots_blink.ino.zip. Flow control is disabled.
> 
> Starting DFU upgrade of type 4, SoftDevice size: 0, bootloader size: 0, application size: 19572
Sending DFU start packet, afterwards we wait for the flash on target to be initialized before continuing.
Timed out waiting for acknowledgement from device.
> 
> Failed to upgrade target. Error is: No data received on serial port. Not able to proceed.
> 
> Possible causes:
> - bootloader, SoftDevice or application on target does not match the requirements in the DFU package.
> - baud rate or flow control is not the same as in the target bootloader.
> Failed to execute script __main__
> - target is not in DFU mode. If using the SDK examples, press Button 4 and RESET and release both to enter DFU mode.

I tried this: 

> - target is not in DFU mode. If using the SDK examples, press Button 4 and RESET and release both to enter DFU mode.

No luck!

**Comment by rwaldron** [comment]
Same thing happens when attempting to upload to the JewelBot (the actual device, not dev board)

**Comment by rwaldron** [comment]
After reading more at https://github.com/NordicSemiconductor/pc-nrfutil, I believe this is a duplicate of #1, so I'm closing

**Comment by SaraJo** [comment]
Are you putting the device in upload mode? (we need a better way to phrase that)

You need to hold down the bottom button for two seconds and it will flash purple to indicate it is ready to upload code.

**Comment by SaraJo** [comment]
Also, this is a different issue, so I reopened.

**Comment by rwaldron** [comment]
> You need to hold down the bottom button for two seconds and it will flash purple to indicate it is ready to upload code.

Ok, I was definitely _not_ doing that, so let me give it a try

**Comment by rwaldron** [comment]
> bottom button 

The one labelled "Action Button"?

Edit: Yes

**Comment by rwaldron** [comment]
Ok, now I get this: 

> Upgrading target on /dev/cu.usbserial-DB00W8D4 with DFU package /private/var/folders/xz/k2lxgs3s43180_tjwn1v_kwc0000gn/T/arduino_build_939294/jewelbots_blink.ino.zip. Flow control is disabled.
> 
> Starting DFU upgrade of type 4, SoftDevice size: 0, bootloader size: 0, application size: 19572
> Sending DFU start packet, afterwards we wait for the flash on target to be initialized before continuing.
> Failed to execute script __main__
> Sending DFU init packet
> Sending firmware file
> DFU upgrade took 17.2695930004s
> Device programmed.
> Failed to execute script __main__

**Comment by SaraJo** [comment]
You're good! That means it went through. 

"failed to execute script main" is a bug listed here. https://github.com/Jewelbots/arduino-library/issues/1

Very confusing for coding, I'd like to tackle that one ASAP.

**Comment by rwaldron** [comment]
@SaraJo that's good news, but the example sketch doesn't seem to work.

**Comment by SaraJo** [comment]
Okay, can you paste the sketch here?

Sara J Chipps
862.201.3065
http://SaraJChipps.com http://sarajchipps.com/

---

No trees were killed in sending this message but a large number of

## electrons were terribly inconvenienced.

Pre-order your Jewelbots http://jewelbots.com/

**Comment by rwaldron** [comment]
```c
#include <Arduino.h>

LED led;

void setup() {
  // Put your setup code here, to run once.
}

void loop() {
  // Put your main code here, to run repeatedly:
  led.on(1, "blue", 650000);
}
```

I see the rainbow display and then the D202 LED settles on green.

**Comment by SaraJo** [comment]
What happens when you unplug it? Same? or, is this on the dev board.

@sm_davison just updated the firmware to not go through the loop until
unplugged, if you have the most recent and the dev board doesn't have the
battery attached that may be hard to test.

These are all things I am adding to the README as we encounter them. Thanks
for being the first eyes.

Sara J Chipps
862.201.3065
http://SaraJChipps.com http://sarajchipps.com/

---

No trees were killed in sending this message but a large number of

## electrons were terribly inconvenienced.

Pre-order your Jewelbots http://jewelbots.com/

**Comment by malwaaare** [comment]
👋🏼   Hey Sara (& everyone else),

I noticed this thread dropped off. I'm having the issue in issue #1 with the `Failed to execute script main` while running the 'hello world' program (code same as pasted above in the thread).

The exception here is that I am booting the Jewelbot into coding mode before uploading. So in between those failure messages, the code seems to compile successfully. However, the light doesn't turn blue.

What might I be missing? I saw a link to the Python library pc-nrfutil, but it wasn't clear how we should make use of that, if at all.

**Comment by SaraJo** [comment]
Hi @mbrewer, have you unplugged your Jewelbot after coding? 

the "Failed to execute script main" is a warning and shouldn't stop the device from accepting your sketch.

**Comment by malwaaare** [comment]
@SaraJo oops—yes, I have! I've been 1) connecting it via USB, 2) booting into coding mode, 3) uploading the code, and 4) unplugging it from USB. And yet, no blue light!

**Comment by SaraJo** [comment]
That's a puzzler. Do you mind updating the firmware to see if that is the issue? Or, have you done that already?

**Comment by malwaaare** [comment]
I don't mind at all! However, I will have to pick this thread back up once I get home -- on my work machine right now w/o my Jewelbot.

**Comment by SaraJo** [comment]
Okay, thank you! Much appreciated.

**Comment by malwaaare** [comment]
So as far as having the firmware 'updated', I'm currently using what the docs recommended.  
![image](https://cloud.githubusercontent.com/assets/4911294/20692596/0b2ed2c2-b5a7-11e6-930a-a5282ac676f2.png)

Is there something else I can update it to?

**Comment by malwaaare** [comment]
Still having an issue with this. I uploaded the above-listed version of the firmware to the Jewelbot (had a blank sketch open + hit 'upload' button). I got a success rainbow.

Booted into coding mode and uploaded the 'hello world' program. Still no blue LED.

**Comment by aulvi** [comment]
@mbrewer These are (roughly) the steps I performed:

1. Turn on the JB, it does a little rainbow dance and buzz to let me know it's alive.
2. Connect up all the USB cables, start the Arduino IDE, place the JB into coding mode by holding The One Button down for two seconds. It should then do a "purple fade" when you let off the button.
3. Create an empty sketch and stick the `LED led; led.on(1, "blue", 2000);` code in the `loop()` function, exactly like the documentation says.
4. Upload the sketch, the Arduino IDE shouldn't barf any errors.
5. Unplug the cable from the JB and the led should light up immediately.

I don't know why, but it took me a couple of tries to get this to work. I ended up adding code for ALL of the LEDs (0-3) while debugging and it seemed fine. Best of luck!
